### ISSUE TITLE
Ability to recompile for each target runtime

### ISSUE THREAD

**Issue by stephentoub** [dotnet/BenchmarkDotNet#1773]
Today in order to compare runtimes, I need to build against a single runtime (the lowest common denominator) with -f and then I can execute against multiple runtimes with --runtimes.  That works in the majority use case, however sometimes you want to answer the question "how does this code compare on different runtimes?", and that includes compile-time impact due to new overloads being introduced, which necessitates building against the appropriate reference assemblies for each version.

**Comment by timcassell** [comment]
Possibly related? It would be nice to be able to recompile targets with different versions of a library/dll. Currently, I can use 2 different versions of the same library in separate projects, both projects referenced by the benchmark project and each called with a different benchmark method, and that compiles just fine, but fails at runtime with `System.Reflection.TargetInvocationException`.

I'm not sure how that could be achieved, considering I have to include the references in the csproj itself.

**Comment by DaveSkender** [comment]
I may have a similar use case that I'm trying to resolve regarding multi-targeting.  Suppose:

```xml
<TargetFrameworks>net6.0;netstandard2.1;</TargetFrameworks>
```

And it produces two DLLs, one for .NET 6.0 and one for .NET Standard 2.1

![image](https://user-images.githubusercontent.com/8432125/144483562-88914316-afb4-4fdc-be51-7d1c1c500013.png)

And now I want to know how they comparatively perform with BenchmarkDotNet, with these attributes?

![image](https://user-images.githubusercontent.com/8432125/144483861-115da0e7-e980-4847-bfe6-c1f8a4472b3f.png)

... I may be confusing the purpose of `RuntimeMoniker`, but I also don't see any other route to do this without creating multiple redundant Benchmark projects.

**Comment by timcassell** [comment]
This is already working for Core runtimes. I sent a PR to fix this for Framework in #2370. I'm also aware of classic Mono not being able to rebuild per target framework, but I'm not sure if we need to add that support. Do you need that, and are there any other runtimes that need this and are not already being rebuilt @stephentoub?

**Comment by stephentoub** [comment]
core and framework are the two I've personally cared about.

**Comment by stephentoub** [comment]
@timcassell, can you clarify what you mean about this already working for Core? I'm not seeing that.  I see a project get build that targets the original .dll and whatever target it was built for.

**Comment by timcassell** [comment]
@stephentoub I added [tests](https://github.com/dotnet/BenchmarkDotNet/blob/914ee7187340761594728497a9f0163a1ecddf67/tests/BenchmarkDotNet.IntegrationTests.ManualRunning.MultipleFrameworks/MultipleFrameworksTest.cs) showing that we are recompiling. Can you share a repro of what you are seeing?

**Comment by stephentoub** [comment]
nuget.config:
```xml
<configuration>
  <packageSources>
    <add key="dotnet8" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json" />
  </packageSources>
</configuration>
```

Program.cs
```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Configs;
using BenchmarkDotNet.Environments;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Running;
using Microsoft.Extensions.Configuration;
using System;

var config = DefaultConfig.Instance
    .AddJob(Job.Default.WithRuntime(CoreRuntime.Core70).WithId(".NET 7.0").WithNuGet("Microsoft.Extensions.Configuration", "7.0.0").AsBaseline())
    .AddJob(Job.Default.WithRuntime(CoreRuntime.Core80).WithId(".NET 8.0").WithNuGet("Microsoft.Extensions.Configuration", "8.0.0-rc.1.23402.3"))
    .AddJob(Job.Default.WithRuntime(CoreRuntime.Core80)
        .WithNuGet("Microsoft.Extensions.Configuration", "8.0.0-rc.1.23402.3")
        .WithNuGet("Microsoft.Extensions.Configuration.Binder", "8.0.0-rc.1.23402.3")
        .WithArguments(new Argument[] { new MsBuildArgument("/p:EnableConfigurationBindingGenerator=true") }));
BenchmarkSwitcher.FromAssembly(typeof(Tests).Assembly).Run(args, config);

[HideColumns("Error", "StdDev", "Median", "RatioSD", "NuGetReferences")]
[MemoryDiagnoser(false)]
public partial class Tests
{
    private MyConfigSection _data = new MyConfigSection();
    private IConfiguration _config;

    [GlobalSetup]
    public void Setup()
    {
        Environment.SetEnvironmentVariable("MyConfigSection__Message", "Hello World!");
        _config = new ConfigurationBuilder()
            .AddEnvironmentVariables()
            .Build();
    }

    [Benchmark]
    public void Load() => _config.Bind("MyConfigSection", _data);

    internal sealed class MyConfigSection
    {
        public string Message { get; set; }
    }
}
```

Benchmarks.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFrameworks>net8.0;net7.0</TargetFrameworks>
    <LangVersion>Preview</LangVersion>
    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
    <ServerGarbageCollection>true</ServerGarbageCollection>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="benchmarkdotnet" Version="0.13.6" />
    <PackageReference Include="Microsoft.Extensions.Configuration" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.EnvironmentVariables" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="8.0.0-rc.1.23402.3" Condition=" '$(TargetFramework)' == 'net8.0' " />
  </ItemGroup>

</Project>
```

Then run with:
```
dotnet run -c Release -f net7.0 --filter **
```

| Method |        Job |  Runtime |                                   Arguments |     Mean | Ratio | Allocated | Alloc Ratio |
|------- |----------- |--------- |-------------------------------------------- |---------:|------:|----------:|------------:|
|   Load |   .NET 7.0 | .NET 7.0 |                                     Default | 2.828 us |  1.00 |    1.3 KB |        1.00 |
|   Load |   .NET 8.0 | .NET 8.0 |                                     Default | 2.097 us |  0.74 |   1.37 KB |        1.05 |
|   Load | Job-UWBTBT | .NET 8.0 | /p:EnableConfigurationBindingGenerator=true | 2.045 us |  0.72 |   1.37 KB |        1.05 |

That last line should be closer to 60ns if it were actually being compiled as desired, and if you add `<EnableConfigurationBindingGenerator>true</EnableConfigurationBindingGenerator>` to the PropertyGroup in the csproj, you'll see it drops to that (but so does the other .NET 8 row).

**Comment by timcassell** [comment]
Oh I see, that's a separate issue. https://github.com/dotnet/BenchmarkDotNet/issues/1377

**Comment by stephentoub** [comment]
> Oh I see, that's a separate issue. https://github.com/dotnet/BenchmarkDotNet/issues/1377

If I `--keepFiles`, I see the supplied argument being passed to the `dotnet build` command in the generated `.bat` file, but that's not the project that needs the argument.

**Comment by timcassell** [comment]
Do you know how to pass it to the proper project? I couldn't figure it out. When I tried it without BDN, it was passed to both projects, but with BDN it doesn't seem to work.

**Comment by stephentoub** [comment]
Nope 😦

**Comment by adamsitnik** [comment]
I commented out all job definitions except the last one and it worked as expected. I assumed that the build issue was caused by the fact that 3 jobs were build in parallel (this is was BDN does by default), and simply the non-source generator net8.0 .dll got consumed by two .NET 8 jobs.

So I modified BDN to disable parallel builds, just to find out it does not solve the problem.

So I've used an old feature: specify build configuration per job via `WithCustomBuildConfiguration`, move all the MSBuild magic to csproj. And it works fine:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFrameworks>net8.0;net7.0</TargetFrameworks>
    <LangVersion>Preview</LangVersion>
    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
    <ServerGarbageCollection>true</ServerGarbageCollection>
    <EnableConfigurationBindingGenerator Condition=" '$(Configuration)' == 'Sourcegen' ">true</EnableConfigurationBindingGenerator>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="benchmarkdotnet" Version="0.13.6" />
    <PackageReference Include="Microsoft.Extensions.Configuration.EnvironmentVariables" Version="7.0.0" Condition=" '$(TargetFramework)' == 'net7.0' "/>
	<PackageReference Include="Microsoft.Extensions.Configuration.EnvironmentVariables" Version="8.0.0-rc.1.23402.3" Condition=" '$(TargetFramework)' == 'net8.0' "/>
    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="8.0.0-rc.1.23402.3" Condition=" '$(TargetFramework)' == 'net8.0' "/>
  </ItemGroup>

</Project>
```

```cs
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Configs;
using BenchmarkDotNet.Environments;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Running;
using Microsoft.Extensions.Configuration;
using System;

var config = DefaultConfig.Instance
    .AddJob(Job.Default.WithRuntime(CoreRuntime.Core70).WithId(".NET 7.0").AsBaseline())
    .AddJob(Job.Default.WithRuntime(CoreRuntime.Core80).WithId(".NET 8.0"))
    .AddJob(Job.Default.WithRuntime(CoreRuntime.Core80)
        .WithId(".NET 8.0 Sourcegen")
        .WithCustomBuildConfiguration("Sourcegen"));

BenchmarkSwitcher.FromAssembly(typeof(Tests).Assembly).Run(args, config);

[HideColumns("Error", "StdDev", "Median", "RatioSD", "NuGetReferences")]
[MemoryDiagnoser(false)]
public partial class Tests
{
    private MyConfigSection _data = new MyConfigSection();
    private IConfiguration _config;

    [GlobalSetup]
    public void Setup()
    {
        Environment.SetEnvironmentVariable("MyConfigSection__Message", "Hello World!");
        _config = new ConfigurationBuilder()
            .AddEnvironmentVariables()
            .Build();
    }

    [Benchmark]
    public void Load() => _config.Bind("MyConfigSection", _data);

    internal sealed class MyConfigSection
    {
        public string Message { get; set; }
    }
}
```

```ini
BenchmarkDotNet v0.13.6, Windows 11 (10.0.22621.1992/22H2/2022Update/SunValley2)
AMD Ryzen Threadripper PRO 3945WX 12-Cores, 1 CPU, 24 logical and 12 physical cores
.NET SDK 8.0.100-preview.6.23330.14
  [Host]             : .NET 7.0.9 (7.0.923.32018), X64 RyuJIT AVX2
  .NET 7.0           : .NET 7.0.9 (7.0.923.32018), X64 RyuJIT AVX2
  .NET 8.0           : .NET 8.0.0 (8.0.23.32907), X64 RyuJIT AVX2
  .NET 8.0 Sourcegen : .NET 8.0.0 (8.0.23.32907), X64 RyuJIT AVX2
```

| Method |                Job |  Runtime | BuildConfiguration |        Mean | Ratio | Allocated | Alloc Ratio |
|------- |------------------- |--------- |------------------- |------------:|------:|----------:|------------:|
|   Load |           .NET 7.0 | .NET 7.0 |            Default | 2,044.77 ns |  1.00 |    1328 B |        1.00 |
|   Load |           .NET 8.0 | .NET 8.0 |            Default | 1,458.20 ns |  0.71 |    1400 B |        1.05 |
|   Load | .NET 8.0 Sourcegen | .NET 8.0 |          Sourcegen |    57.50 ns |  0.03 |     112 B |        0.08 |

**Comment by timcassell** [comment]
> I commented out all job definitions except the last one and it worked as expected.

Really? I tried that and got the same result as the job without the MsBuildArgument.

**Comment by timcassell** [comment]
I see the issue. On each build, by default the binaries are placed in `./bin/<Configuration>/<Runtime>`, so with 2 builds using the same configuration and runtime, 1 of them will overwrite the other. Using a new configuration places the binaries in a new directory so it doesn't overwrite the other build.

Possible solutions:
- Only build before the benchmarks are ran for each exe.
- Move the binaries after each build (parallel builds will have to be disabled).
- Build the source proj separately with manually specifying the binaries output path, then build the auto generated csproj separately, referencing the source binaries.

**Comment by timcassell** [comment]
I sent a fix #2393 which makes the host runtime no longer matter (unless you need classic Mono or in-process).
